contextualized_topic_models/models/ctm.py

```diff
@@ -379,44 +506,80 @@ def get_doc_topic_distribution(self, dataset, n_samples=20):
         self.model.eval()
 
         loader = DataLoader(
-            dataset, batch_size=self.batch_size, shuffle=False,
-            num_workers=self.num_data_loader_workers)
-        pbar = tqdm(n_samples, position=0, leave=True)
+            dataset,
+            batch_size=self.batch_size,
+            shuffle=False,
+            num_workers=self.num_data_loader_workers,
+        )
         final_thetas = []
-        for sample_index in range(n_samples):
-            with torch.no_grad():
-                collect_theta = []
-
-                for batch_samples in loader:
-                    # batch_size x vocab_size
-                    X_bow = batch_samples['X_bow']
-                    X_bow = X_bow.reshape(X_bow.shape[0], -1)
-                    X_contextual = batch_samples['X_contextual']
-
-                    if "labels" in batch_samples.keys():
-                        labels = batch_samples["labels"]
-                        labels = labels.to(self.device)
-                        labels = labels.reshape(labels.shape[0], -1)
-                    else:
-                        labels = None
-
-                    if self.USE_CUDA:
-                        X_bow = X_bow.cuda()
-                        X_contextual = X_contextual.cuda()
-
-                    # forward pass
-                    self.model.zero_grad()
-                    collect_theta.extend(self.model.get_theta(X_bow, X_contextual, labels).cpu().numpy().tolist())
-
-                pbar.update(1)
-                pbar.set_description("Sampling: [{}/{}]".format(sample_index + 1, n_samples))
-
-                final_thetas.append(np.array(collect_theta))
-        pbar.close()
-        return np.sum(final_thetas, axis=0) / n_samples
+        with torch.no_grad():
+            for batch_samples in tqdm(loader):
+                # batch_size x vocab_size
+                X_bow = batch_samples["X_bow"]
+                X_bow = X_bow.reshape(X_bow.shape[0], -1)
+                X_contextual = batch_samples["X_contextual"]
+
+                if "labels" in batch_samples.keys():
+                    labels = batch_samples["labels"]
+                    labels = labels.to(self.device)
+                    labels = labels.reshape(labels.shape[0], -1)
+                else:
+                    labels = None
+
+                if self.USE_CUDA:
+                    X_bow = X_bow.cuda()
+                    X_contextual = X_contextual.cuda()
+
+                # forward pass
+                self.model.zero_grad()
+                mu, log_sig = self.model.get_posterior(X_bow, X_contextual, labels)
+                thetas = self.model.sample(mu, log_sig, n_samples).cpu().numpy()
+                final_thetas.append(thetas)
+        return np.concatenate(final_thetas, axis=0)
+
+    def get_doc_topic_distribution_iterator(self, dataset, n_samples=20):
```

Comment:
When is this supposed to be called?

Reply:
`get_doc_topic_distribution_iterator` enables _significant_ memory reduction on large datasets, as we no longer need to collect `theta` values for all examples in our dataset. 

The old `get_doc_topic_distribution` had memory usage O(N * T * K), where N is the number of examples in the dataset, T is the number of topics, and K is the number of samples. 

After my efficient `sample` function, the `get_doc_topic_distribution` function goes down to O(N * T + B * T * K), with B being the batch size (MUCH smaller than N), where now the K samples are only computed for a batch. 

Finally, we can achieve O(B * T * K) by avoiding building that final `return np.concatenate(final_thetas, axis=0)` and the corresponding list `final_thetas` by yielding each batch's topics as they are generated. This is the same concept as how in Python 3 `range(10_000)` does not need to generate 10,000 integers in order for the loop to work, we just generate them one at a time as needed with a generator.

Let me know if you have any other questions!